tabular/src/autogluon/tabular/models/ensemble/bagged_ensemble_model.py

```diff
@@ -597,3 +597,71 @@ def _construct_empty_oof(self, X, y):
             oof_pred_proba = np.zeros(shape=len(X), dtype=np.float32)
         oof_pred_model_repeats = np.zeros(shape=len(X), dtype=np.uint8)
         return oof_pred_proba, oof_pred_model_repeats
+
+    def _preprocess_fit_resources(self, kwargs):
+        """Pass along to child models to avoid altering up-front"""
+        return kwargs
+
+    # TODO: Currently double disk usage, saving model in HPO and also saving model in bag
+    def _hyperparameter_tune(self, X_train, y_train, k_fold, scheduler_options=None, preprocess_kwargs=None, **kwargs):
+        if len(self.models) != 0:
+            raise ValueError('self.models must be empty to call hyperparameter_tune, value: %s' % self.models)
+
+        self.model_base.feature_metadata = self.feature_metadata  # TODO: Move this
+
+        # TODO: Preprocess data here instead of repeatedly
+        if preprocess_kwargs is None:
+            preprocess_kwargs = dict()
+        X_train = self.preprocess(X=X_train, preprocess=False, fit=True, **preprocess_kwargs)
+        kfolds = generate_kfold(X=X_train, y=y_train, n_splits=k_fold, stratified=self.is_stratified(), random_state=self._random_state, n_repeats=1)
+
+        train_index, test_index = kfolds[0]
+        X_train_fold, X_val_fold = X_train.iloc[train_index, :], X_train.iloc[test_index, :]
+        y_train_fold, y_val_fold = y_train.iloc[train_index], y_train.iloc[test_index]
+        orig_time = scheduler_options[1]['time_out']
+        scheduler_options[1]['time_out'] = orig_time * 0.8  # TODO: Scheduler doesn't early stop on final model, this is a safety net. Scheduler should be updated to early stop
+        hpo_models, hpo_model_performances, hpo_results = self.model_base.hyperparameter_tune(X_train=X_train_fold, y_train=y_train_fold, X_val=X_val_fold, y_val=y_val_fold, scheduler_options=scheduler_options, **kwargs)
+        scheduler_options[1]['time_out'] = orig_time
+
+        bags = {}
+        bags_performance = {}
+        for i, (model_name, model_path) in enumerate(hpo_models.items()):
+            child: AbstractModel = self._child_type.load(path=model_path)
+            y_pred_proba = child.predict_proba(X_val_fold)
+
+            # TODO: Create new Ensemble Here
+            bag = copy.deepcopy(self)
+            bag.name = bag.name + os.path.sep + str(i)
+            bag.set_contexts(self.path_root + bag.name + os.path.sep)
+
+            oof_pred_proba, oof_pred_model_repeats = self._construct_empty_oof(X=X_train, y=y_train)
+            oof_pred_proba[test_index] += y_pred_proba
+            oof_pred_model_repeats[test_index] += 1
```

Comment:
This might not work correctly if AUC is the optimization metric since AUC cares only about rank. [Here](https://scikit-learn.org/stable/auto_examples/model_selection/plot_roc_crossval.html) are some ideas on how to better handle it.
I don't expect this to be fixed in this PR, but let's add some TODO to address it later.

Reply:
This is already a TODO at top of Trainer (I've known about it for a long time)